src/Commands/Runner.h

```diff
@@ -534,6 +566,35 @@ class Runner {
 
   {}
 
+#ifdef WITH_REPLXX
+  void CompletionHook(char const* pref, int ctx, replxx_completions* lc) {
+    std::string prefix(pref);
+    auto spacePos = prefix.find_first_of(" \t");
+    auto subCmdPos = prefix.find_first_not_of(" \t", spacePos);
+    for (const auto& c : _commands) {
+      const std::string& cname = c.first;
+      if (!cname.compare(0, prefix.size(), prefix)) {
+        replxx_add_completion(lc, cname.c_str());
+        continue;
+      }
+      if (spacePos == std::string::npos) {
```

Comment:
It just occurred to me that I am not clear that this does what is intended based on what has been typed so far.  Suppose, for example, that the user wants to run "list leaked":

I would expect that if the user has already typed "l" that tab completion would just allow "list" but if the line currently contains "list l" that tab completion should allow that second "l" to be expanded to "leaked" but not to "list" because "list list" doesn't make any sense.  Will the current form of the code allow that second "l" to be expanded to "list"?

Another minor wrinkle here is that chap currently ignores any (admittedly unexpected) white space at the beginning of the line, so "   list  leaked", although admittedly unexpected, is equivalent to "list leaked".  With that in mind, leading white space probably shouldn't influence tab completion.

Reply:
> It just occurred to me that I am not clear that this does what is intended based on what has been typed so far. Suppose, for example, that the user wants to run "list leaked":
 
> I would expect that if the user has already typed "l" that tab completion would just allow "list" but if the line currently contains "list l" that tab completion should allow that second "l" to be expanded to "leaked" but not to "list" because "list list" doesn't make any sense. Will the current form of the code allow that second "l" to be expanded to "list"?

This is exactly how it works at the moment. The first 'l' will be expanded to list while the second will expand to leaked. I just gave it a try.  Pressing `l + TAB l +TAB` results in `list leaked`

> Another minor wrinkle here is that chap currently ignores any (admittedly unexpected) white space at the beginning of the line, so " list leaked", although admittedly unexpected, is equivalent to "list leaked". With that in mind, leading white space probably shouldn't influence tab completion.

Yes, I agree. I have been thinking about it as well. I can add it trivially.